numpy/core/src/common/npy_argparse.h

```diff
@@ -0,0 +1,151 @@
+ *
+ * Internally CPython achieves similar things by using a code generator
+ * argument clinic. NumPy may well decide to use argument clinic or a different
+ * solution in the future.
+ */
+
+NPY_NO_EXPORT int
+PyArray_PythonPyIntFromInt(PyObject *obj, int *value);
+
+
+#define _NPY_MAX_KWARGS 15
+
+typedef struct {
+    int npositional;
+    int nargs;
+    int npositional_only;
+    int nrequired;
+    int nkwarg;
+    // char intent[MAX_KWARGS];
+    /* Null terminated list of keyword argument name strings */
+    PyObject *kw_strings[_NPY_MAX_KWARGS];
+} _NpyArgParserCache;
+
+
+/*
+ * Macros to help support `METH_FASTCALL` when available and fall back to
+ * normal parsing when not.
+ * The signature of function is different in the two cases, so this switches.
+ */
+#ifdef METH_FASTCALL
+
+#if PY_VERSION_HEX >= 0x03070000
+    #define NPY_METH_KWARGS_FASTCALL METH_FASTCALL | METH_KEYWORDS
+#else
+    /* Python 3.6 _officially_ did not support METH_FASTCALL, but it does. */
+    #define NPY_METH_KWARGS_FASTCALL METH_FASTCALL
+#endif
+#define NPY_PARAMS_DEF \
+    PyObject *const *args, Py_ssize_t len_args, PyObject *kwnames
+#define NPY_PARAMS_PASS args, len_args, kwnames
+#define NPY_PARAMS_GET_ARG(i) args[i]
+#define NPY_PARAMS_DEFINE_LEN_ARGS do {} while(0)
+#define NPY_PARAMS_KWARG_OBJ kwnames
+
+#else  /* METH_FASTCALL */
+
+#define NPY_METH_KWARGS_FASTCALL METH_VARARGS | METH_KEYWORDS
+#define NPY_PARAMS_DEF PyObject *args, PyObject *kwargs
+#define NPY_PARAMS_PASS args, kwargs
+#define NPY_PARAMS_GET_ARG(i) PyTuple_GET_ITEM(args, i)
+/* Allow args == NULL, for when all args are already caught by `*args` */
+#define NPY_PARAMS_DEFINE_LEN_ARGS Py_ssize_t len_args = \
+        (args == NULL) ? 0 : PyTuple_GET_SIZE(args)
+#define NPY_PARAMS_KWARG_OBJ kwargs
+
+#endif  /* METH_FASTCALL */
+
+
+#define NPY_PREPARE_ARGPARSER \
+        static _NpyArgParserCache __argparse_cache = {-1};      \
+
+
+/**
+ * Macro to help with argument parsing (together with above definitions)
+ *
+ * The pattern for using this macro is by defining the method as:
+ *
+ * @code
+ * static PyObject *my_method(PyObject *self, NPY_ARGUMENTS_DEF)
+ * {
+ *     NPY_PREPARE_ARGPARSER;
+ *
+ *     PyObject *argument1;
+ *     int argument2 = -1;
+ *     if (!npy_parse_arguments("method", 1, -1, NPY_ARGUMENTS_PASS),
+ *                "argument1", NULL, &argument1,
+ *                "argument2", &PyArray_PythonPyIntFromInt, &argument2,
+ *                NULL, NULL, NULL) {
+ *          return NULL;
+ *      }
+ * }
+ * @endcode
+ *
+ * The macros ``NPY_PARAMS_GET_ARG`` to get a specific positional argument and
+ * ``NPY_PARAMS_DEFINE_LEN_ARGS`` to define len_args as the number of positional
+ * arguments may be helpful in some cases.
+ *
+ * @param funcname
+ * @param nrequired Number of required arguments
+ * @param npositional Number of positional only arguments, can be -1 to
+ *                    indicate that all arguments may be passed positionally.
```

Comment:
Not sure I understand this - do you mean _must_ be passed positionally?

Or are you trying to describe "not keyword-only"?

Reply:
The latter, not keyword only, which seems simpler than to count the number of keyword only. Will try to rephrase if this seems to go anywhere.